crypto/fipsmodule/ec/ecp_nistz.inl

```diff
@@ -0,0 +1,449 @@
+/* Copyright (c) 2014, Intel Corporation.
+ *
+ * Permission to use, copy, modify, and/or distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
+ * SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
+ * OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+ * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE. */
+
+/* Developers and authors:
+ * Shay Gueron (1, 2), and Vlad Krasnov (1)
+ * (1) Intel Corporation, Israel Development Center
+ * (2) University of Haifa
+ * Reference:
+ *   Shay Gueron and Vlad Krasnov
+ *   "Fast Prime Field Elliptic Curve Cryptography with 256 Bit Primes"
+ *   http://eprint.iacr.org/2013/816 */
+#include "ecp_nistz.h"
+
+#if defined(__GNUC__) || defined(__clang__)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wsign-conversion"
+#endif
+
+#define RENAME_FUNC(prefix, bits, func) prefix ## bits ## _ ## func
+
+#define point_add(prefix, bits) RENAME_FUNC(prefix, bits, point_add)
+#define point_double(prefix, bits) RENAME_FUNC(prefix, bits, point_double)
+#define point_mul(prefix, bits) RENAME_FUNC(prefix, bits, point_mul)
+
+typedef Limb Elem[FE_LIMBS];
+typedef Limb ScalarMont[FE_LIMBS];
+typedef Limb Scalar[FE_LIMBS];
+
+typedef struct {
+  Limb X[FE_LIMBS];
+  Limb Y[FE_LIMBS];
+  Limb Z[FE_LIMBS];
+} NIST_POINT;
+
+typedef struct {
+  Limb X[FE_LIMBS];
+  Limb Y[FE_LIMBS];
+} NIST_POINT_AFFINE;
+
+#define TBL_SZ (1 << (W_BITS - 1))
+#define W_MASK ((1 << (W_BITS + 1)) - 1)
+
+static inline Limb is_equal(const Elem a, const Elem b) {
+  return LIMBS_equal(a, b, FE_LIMBS);
+}
+
+static inline Limb is_zero(const BN_ULONG a[FE_LIMBS]) {
+  return LIMBS_are_zero(a, FE_LIMBS);
+}
+
+static inline void copy_conditional(Elem r, const Elem a,
+                                                const Limb condition) {
+  for (size_t i = 0; i < FE_LIMBS; ++i) {
+    r[i] = constant_time_select_w(condition, a[i], r[i]);
+  }
+}
+
+static inline void elem_add(Elem r, const Elem a, const Elem b) {
+  LIMBS_add_mod(r, a, b, Q, FE_LIMBS);
+}
```

Comment:
It's annoying, but it's better for us to keep the stuff that was from gfp_p384.c (that I wrote) in a separate file from the stuff that was originally from Intel so that when we move things to Rust, we can keep track of the provinence correctly to make sure the license stuff is done correctly. So please split gfp_p384.c into `gfp.inl` and `gfp_p384.c`. (In the Rust equivalent to this, I will probably not end up deriving the code from Intel-sourced code, and I don't want to muddy things.)

Making this change will also (I hope) make it so GitHub will show my the actual diffs between the original files and the new files, which would make this easier to review.